src/wallet/wallet.cpp

```diff
@@ -4997,125 +5086,121 @@ void CWallet::GetFilteredNotes(
+    filters.ignoreSpent = ignoreSpent;
+    filters.requireSpendingKey = requireSpendingKey;
+    filters.ignoreLocked = ignoreLocked;
 
-        // Filter coinbase transactions that don't have Sapling outputs
-        if (wtx.IsCoinBase() && wtx.mapSaplingNoteData.empty()) {
-            continue;
-        }
+    auto instance = NotesHelperIndex::getInstance();
 
-        for (auto & pair : wtx.mapSproutNoteData) {
-            JSOutPoint jsop = pair.first;
-            SproutNoteData nd = pair.second;
-            SproutPaymentAddress pa = nd.address;
+    // default: all available wallet addresses when filterAddresses is empty
+    if (filterAddresses.size() < 1) {
+        std::set<libzcash::SproutPaymentAddress> sproutAddrs;
+        GetSproutPaymentAddresses(sproutAddrs);
 
-            // skip notes which belong to a different payment address in the wallet
-            if (!(filterAddresses.empty() || filterAddresses.count(pa))) {
-                continue;
-            }
+        std::set<libzcash::SaplingPaymentAddress> SaplingAddrs;
+        GetSaplingPaymentAddresses(SaplingAddrs);
 
-            // skip note which has been spent
-            if (ignoreSpent && nd.nullifier && IsSproutSpent(*nd.nullifier)) {
-                continue;
-            }
+        if (sproutAddrs.size() > 0)
+            filterAddresses.insert(sproutAddrs.begin(), sproutAddrs.end());
+        if (SaplingAddrs.size() > 0)
+            filterAddresses.insert(SaplingAddrs.begin(), SaplingAddrs.end());
 
-            // skip notes which cannot be spent
-            if (requireSpendingKey && !HaveSproutSpendingKey(pa)) {
-                continue;
-            }
+        if (filterAddresses.size() < 1)
+            return;
+    }
 
-            // skip locked notes
-            if (ignoreLocked && IsLockedNote(jsop)) {
-                continue;
-            }
+    auto& index = instance->index.get<by_timestamp>();
 
-            int i = jsop.js; // Index into CTransaction.vJoinSplit
-            int j = jsop.n; // Index into JSDescription.ciphertexts
+    for (auto& zaddr : filterAddresses) {
+        uint256 paymentaddress_hash;
+        if (boost::get<libzcash::SproutPaymentAddress>(&zaddr) != nullptr) {
+            SproutPaymentAddress paymentaddress = *boost::get<libzcash::SproutPaymentAddress>(&zaddr);
+            paymentaddress_hash = paymentaddress.GetHash();
 
-            // Get cached decryptor
-            ZCNoteDecryption decryptor;
-            if (!GetNoteDecryptor(pa, decryptor)) {
-                // Note decryptors are created when the wallet is loaded, so it should always exist
-                throw std::runtime_error(strprintf("Could not find note decryptor for payment address %s", keyIO.EncodePaymentAddress(pa)));
-            }
+            auto itr_sprout = GetNotesByType(paymentaddress_hash, NoteType::sprout, timestamp);
+            auto itr = itr_sprout.begin();
+
+            while (itr != itr_sprout.end()) {
```

Comment:
Use a `for` loop.

Reply:
https://github.com/zcash/zcash/pull/4399/commits/743d34166706c32a56225bd1102d41c03a82c517